wallet/src/actors/worker/methods.rs

```diff
@@ -283,4 +297,175 @@ impl Worker {
 
         Ok(signed_data)
     }
+
+    /// Try to synchronize the information for a wallet to whatever the world state is in a Witnet
+    /// chain.
+    pub fn sync(
+        &self,
+        wallet_id: &str,
+        wallet: types::SessionWallet,
+        since_epoch: u32,
+    ) -> Result<()> {
+        log::info!("Trying to start wallet synchronization");
+        // TODO: read the limit from configuration
+        let limit = 100i64;
+        let mut latest_epoch = since_epoch;
+        let mut since_epoch = i64::from(since_epoch);
+
+        loop {
+            // Ask a Witnet node for epochs and ids for all the blocks that we have been missing
+            let get_block_chain_future = self.get_block_chain(since_epoch, limit);
+            let block_chain = futures03::executor::block_on(get_block_chain_future)?;
+            log::debug!("Received chain: {:?}", block_chain);
+
+            // For each of the blocks we have been informed about, ask a Witnet node for its contents
+            for ChainEntry(epoch, id) in block_chain {
+                let get_block_future = self.get_block(id);
+                let block = futures03::executor::block_on(get_block_future)?;
+                log::debug!("For epoch {}, got block: {:?}", epoch, block);
+
+                // Process each block
+                futures03::executor::block_on(self.handle_block(block, wallet_id, wallet.clone()))?;
+                latest_epoch = epoch;
+            }
+
+            // Persist the epoch of the last synced block. Doing this by batches instead of by block
+            // should be safe under the assumption that the same transaction cannot be processed
+            // twice by a single wallet.
+            wallet.update_last_sync(latest_epoch)?;
```

Comment:
This assumption needs to be validated.

Reply:
I think the `handle_block` method should save the hash of the last handled block, and then when handling a new block do a quick check that `block_beacon.hash_prev_block` is equal to that hash. This way we avoid any problems with missing blocks, duplicated blocks, forked blocks, and blocks being processed out of order.

Reply:
Great suggestion. It works like that now :ok_hand: